### ISSUE TITLE
Signficantly more GPU consumption when training DALLE with DiscreteVAE vs OpenAI VAE

### ISSUE THREAD

**Issue by richcmwang** [lucidrains/DALLE-pytorch#233]
I notice a significantly more GPU consumption (2x-3x) when training DALLE using DiscreteVAE vs OpenAI VAE.  The `train_dalle.py` I tested is an earlier version I downloaded around 4 weeks ago - I notice some update, but not sure it addresses the issue.  

The overhead happens at `loss = dalle(text, images, mask=mask, return_loss=True)` in the `train_dalle.py`.  Specifically, inside the `forward` in the class DALLE, after these two lines:
`out = self.transformer(tokens, mask=mask)`  
`loss_img = F.cross_entropy(
            logits[:, :, self.text_seq_len:], labels[:, self.text_seq_len:], ignore_index=0)` 
            
Each line doubles the memory usage with DiscreteVAE whereas OpenAI VAE only adds a small amount of usage.  I was wondering why?  In addition, Do we need to freeze the VAE model in the `forward` (probably not necessary because no_grad in the `get_codebook_indices`)?

**Comment by lucidrains** [comment]
@richcmwang Hi Rich! Although I'm not entirely sure if it'll fix the problem, do you want to try https://github.com/lucidrains/DALLE-pytorch/commit/19f42124458b575305a1a386d0f88cb22790a6d5 ?

**Comment by richcmwang** [comment]
Hi Phil, I realize it's the model output difference: Inside the `forward` function:
`out = self.transformer(tokens, mask=mask)`
`logits = self.to_logits(out)`
DIscretVAE: out.shape: (4, 4352, 512) logits.shape (4, 4352, 57600)
OpenAI VAE: out.shape: (4, 1280, 512), logits.shape (4, 1280, 57600)
It seems to be a design issue.  The DiscreteVAE model size is much smaller, but the output tensor is x4 bigger.  Is it possible to reduce the output tensor while retaining the performance?

**Comment by lucidrains** [comment]
ohh got it, yes you'll need to make sure your discrete vae has at least 3 layers, so that it ends up with a 32x32 feature map for images that are 256px

**Comment by richcmwang** [comment]
Got it thanks!
